include/Factory.h

```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include "macros.h"
+
+class Piece;
+
+template<typename T>
+class Factory {
+
+    static auto &getMap() {
+        static std::map<char, pFnc> m_map;
+        return m_map;
+    }
+
+public:
+    using pFnc = std::unique_ptr<T>(*)();
+
+    static unique_ptr<T> create(const char &name);
+
+    static bool registerIt(const char &name, pFnc f);
+
+};
+
+template<typename T>
+unique_ptr<T> Factory<T>::create(const char &name) {
+    auto it = Factory<T>::getMap().find(name);
+    if (it == Factory<T>::getMap().end())
```

Comment:
Calling getMap() twice looke like a waste. It can be easily spared to only one call to the function